content/en/docs/concepts/extend-kubernetes/compute-storage-net/resource-drivers.md

```diff
@@ -0,0 +1,278 @@
+---
+title: DRA Resource Drivers
+description: Resource drivers provide non-trivial allocation logic and management for devices or resources that require vendor-specific or just complex setup, such as GPUs, NICs, FPGAs, etc.
+content_type: concept
+weight: 10
+---
+
+<!-- overview -->
+{{< feature-state for_k8s_version="v1.26" state="alpha" >}}
+
+Kubernetes provides a
+[Dynamic Resource Allocation (DRA)](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation)
+that can be leveraged to provide more complex hardware resources to workloads with custom resource
+accounting.
+
+Similarly to {{< glossary_tooltip term_id="device-plugin" text="Device Plugins">}}, instead of
+customizing the code for Kubernetes itself, vendors can implement a resource driver that you deploy
+into the cluster to account for and control the allocation of GPUs, high-performance NICs, FPGAs,
+InfiniBand adapters, and other similar computing resources that may require vendor specific
+initialization and setup.
+
+With device plugins, the scheduler was given a trivial, numerical representation of the resources
+available on node for consideration during scheduling as an extended resource.
+
+With DRA, the scheduler is offloading the task of allocating and accounting for non-native resources
+to the resource driver, which manages such resources in the cluster.
+
+Resource driver consists of two main components:
+
+- controller (one per cluster), manages hardware resources allocation for
+  {{< glossary_tooltip term_id="resource-claim" text="Resource Claims">}}
+- kubelet-plugin (one per node with resource), that:
+  - discovers the supported hardware
+  - announces discovered hardware to the Resource Driver controller
+  - prepares the hardware allocated for Pod when Pod comes to the Kubelet
+  - unprepares the hardware allocated for Pod when Pod has reached final state or is being deleted.
+
+Communications between controller and kubelet-plugin can be done in two ways:
+
+- through Custom Resource Definition objects provided by and designed for the resource driver
+- through ResourceHandle which is a part of AllocationResult provided by the controller in case of
+  successful allocation
+
+General recommendations:
+
+- resource driver name pattern: `<HW type>.resource.<companyname>.<companydomain>`, e.g.
```

Comment:
```suggestion
- A resource driver name should follow the following pattern: `<HW type>.resource.<companyname>.<companydomain>`, e.g.
```
Is this a "SHOULD" or a "MUST"?

Reply:
We don't restrict or enforce this anyhow, it's just something that we've agreed between who is involved in DRA at the moment. One can create the driver with complete different pattern name, this is just a suggestion.

Reply:
However, we (Kubernetes) should document expectations here, especially around the final DNS domain part. What companies do with the part before their company name we mind rather less about.

To me `gpu.resource.example.com` and `fpga.foo.example.net` are fine as examples, as is `expansion-j5.wopr.norad.mil`, but not `cardpunch.ibm-029.some-random-integrator-company`.

Reply:
Is it the length of the resource-driver name or what in particular is wrong with the last one? IMHO we don't care about the resource-driver name at all, It is almost invisible to users. ResourceClass name is what's referenced in ResourceClaims, not resource driver name. Wasn't main point of this recommendation to prevent conflicts in API object names. If, for example, all GPU vendors would name their resources as `gpu.resource` or `mydevice.gpu` or anything non vendor specific, there would be a chance for name collision.

Reply:
> If, for example, all GPU vendors would name their resources as gpu.resource or mydevice.gpu or anything non vendor specific, there would be a chance for name collision.

If `cardpunch.ibm-029.some-random-integrator-company` is bad, `some-random-integrator-company.ibm-029.cardpunch` is worse. More chance of collision.

The problem with the last one in my examples is that `some-random-integrator-company` is not a valid DNS name. If we implicitly or tacitly permit these, we're opening end users up to that collision risk.

Even if the implementation code in Kubernetes doesn't check or care, interoperability works better if we strongly recommend that companies use their own actual DNS name as the last part here. That way, when someone does a foolish thing, we can point to our advice and highlight the good practice.
Also, people who want to can write an admission controller that warns on poorly chosen resource driver strings.

Reply:
> > If, for example, all GPU vendors would name their resources as gpu.resource or mydevice.gpu or anything non vendor specific, there would be a chance for name collision.z
> 
> If `cardpunch.ibm-029.some-random-integrator-company` is bad, `some-random-integrator-company.ibm-029.cardpunch` is worse. More chance of collision.

Yes, that was exactly my point.

> The problem with the last one in my examples is that `some-random-integrator-company` is not a valid DNS name. If we implicitly or tacitly permit these, we're opening end users up to that collision risk.
> 
> Even if the implementation code in Kubernetes doesn't check or care, interoperability works better if we strongly recommend that companies use their own actual DNS name as the last part here. That way, when someone does a foolish thing, we can point to our advice and highlight the good practice. Also, people who want to can write an admission controller that warns on poorly chosen resource driver strings.

We cannot possibly demand actual DNS because we expect resource drivers to be created by community as well as hardware vendors. Resource driver names without actual existing company domain name are inevitable.

In addition there will be drivers that span across hardware vendors. For example, a resource driver that would allocate jointly one piece of hardware F1 from FOO INC with one piece of hardware B1 from BAR INC as a bundle. Such resource driver will need a name too.

Reply:
Community projects do have domain names. This one has several.

Let's get SIG Architecture to confirm the view here.